### PR TITLE
Fixing the current state update when lncli debuglevel and lncli setmccfg are called

### LINKED ISSUE #8793: [bug]: `lncli getdebuginfo` not showing the current state of config
I used `lncli debuglevel` to change the debuglevel and `lncli setmccfg` to change the estimator. In both cases the config under lncli getdebuginfo has not changed. It seems that only the config at the start time is displayed there and not the current state.

### PR DESCRIPTION
**Pull request by MPins**
Fixes #8793	

modified:   rpcserver.go
modified:   server.go
 modified:   routing/missioncontrol.go

## Change Description
Changed the DebugLevel function to start updating the main cfg DebugLevel variable.

Created the function UpdateRoutingConfig to just update the main cfg  values, this function is a callback function propagated by NewMissionControl function called when creating a new server on server.go. And finally, call the callback function on SetConfig on routing/missioncontrol.go

## Steps to Test
After running LND with the wallet unlocked, send the lncli commands to change the debuglevel and estimator and check if they are updated unsing 'getdebuginfo'.

### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [ ] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [ ] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [ ]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -923,8 +923,11 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
 				routingConfig.ProbabilityEstimatorType)
 		}
 	}
-
+	// UpdateEstimatorValue is a function that will be called by the
+	// mission control instance every time a new estimator stratefu
```
Comment (alexbosworth) [inline server.go:927]: stratefu?

File: routing/missioncontrol.go [hunk 2]
```diff
@@ -121,6 +121,10 @@ type MissionControl struct {
 // MissionControlConfig defines parameters that control mission control
 // behaviour.
 type MissionControlConfig struct {
+	// callback is a function that is called whenever the mission control
```
Comment (Abdulkbk) [inline routing/missioncontrol.go:124]: nit: UpdateEstimator is a ... instead of "callback" in the comment

File: rpcserver.go [hunk 3]
```diff
@@ -6901,6 +6901,16 @@ func (r *rpcServer) DebugLevel(ctx context.Context,
 	err := build.ParseAndSetDebugLevels(req.LevelSpec, r.cfg.LogWriter)
 	if err != nil {
 		return nil, err
+	} else {
```
Comment (Abdulkbk) [inline rpcserver.go:6904]: nit: You can omit the else block and write the statement within it after the if block, since we return early if there's an error.

File: rpcserver.go [hunk 4]
```diff
@@ -6901,6 +6901,16 @@ func (r *rpcServer) DebugLevel(ctx context.Context,
...
+		// There is no error parsing the LevelSpec, so we'll update
+		// the current debug level, without checking everything again
+		levels := strings.Split(req.LevelSpec, ",")
+		// If the first entry has no =, treat it as the log level for
+		// all subsystems and the debuglevel is updated
+		if !strings.Contains(levels[0], "=") {
```
Comment (Abdulkbk) [inline rpcserver.go:6910]: Clever way of distinguishing between sub-system level and individual sub-system levels config :)

**Comment by Abdulkbk** [comment]
Nice work @MPins. I left a few comments. 

Also, I don't think this [commit](https://github.com/lightningnetwork/lnd/pull/8857/commits/31816914381f1002d274a47178c982b22410bafe) is necessary since all you did was add comments.

**Comment by MPins** [comment]
> Nice work @MPins. I left a few comments.
> 
> Also, I don't think this [commit](https://github.com/lightningnetwork/lnd/pull/8857/commits/31816914381f1002d274a47178c982b22410bafe) is necessary since all you did was add comments.

That's weird ... because I've added the line containing cfg.UpdateEstimatorValue(cfg.Estimator) also

File: rpcserver.go (hunk 3 above)
Comment (MPins) [reply rpcserver.go:6904]: Done ... thanks!

File: routing/missioncontrol.go (hunk 2 above)
Comment (MPins) [reply routing/missioncontrol.go:124]: I changed the name of the function and did not update the comment ... just fixed

File: server.go (hunk 1 above)
Comment (MPins) [reply server.go:927]: Oops ... just fixed the comment

File: server.go [hunk 5]
```diff
@@ -1680,6 +1683,12 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
 	return s, nil
 }
 
+// UpdateEstimatorValue is a callback function to change estimator value in the  
+// main cfg when the lncli setmccfg is called to change the estimator value. 
+ func (s *server) UpdateEstimatorValue(estimator routing.Estimator) {
+	s.cfg.Estimator = estimator
```
Comment (guggero) [inline server.go:1689]: You're just updating the estimator interface here, which is just used as a cache after parsing the config values into the correct structure.
Try something like this instead (untested, but you should get the idea):
```
	switch c := estimator.Config().(type) {
	case *routing.AprioriConfig:
		targetCfg := s.cfg.SubRPCServers.RouterRPC.AprioriConfig
		targetCfg.PenaltyHalfLife = c.PenaltyHalfLife
		targetCfg.Weight = c.AprioriWeight
		targetCfg.CapacityFraction = c.CapacityFraction
		targetCfg.HopProbability = c.AprioriHopProbability

	case *routing.BimodalConfig:
		targetCfg := s.cfg.SubRPCServers.RouterRPC.BimodalConfig
		targetCfg.Scale = int64(c.BimodalScaleMsat)
		targetCfg.NodeWeight = c.BimodalNodeWeight
		targetCfg.DecayTime = c.BimodalDecayTime
	}
```

**Comment by MPins** [comment]
Hello everyone, the code is working as expected! But the changes made with `lncli setmccfg` are not impacting the information we get from `lncli getdebuginfo' . It looks like the GetDebugInfo function is getting this information from somewhere else.
https://github.com/lightningnetwork/lnd/blob/71ba355d903f01990718866c358c2ccac9402891/rpcserver.go#L3164

**Comment by Chinwendu20** [comment]
So `GetDebugInfo` is getting its info from a different config from the one that you are updating (`r.cfg`)
https://github.com/lightningnetwork/lnd/blob/b67af4a7061f5d14e0f537b8073afa3101d8d16b/rpcserver.go#L3166

You would have to change it to `r.server.cfg` to get the updated one.

Then maybe change this from `r.cfg.LogWriter` to `r.server.cfg.LogWriter` so that all the config updates can be in one place:
https://github.com/lightningnetwork/lnd/blob/b67af4a7061f5d14e0f537b8073afa3101d8d16b/rpcserver.go#L6901

I think maybe the config field in the `rpcserver` struct should be removed entirely since we can access config from the `server` field but I would like hear what other ppl think about it.

File: server.go [hunk 6]
```diff
@@ -1686,7 +1686,20 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
 // UpdateEstimatorValue is a callback function to change estimator value in the  
 // main cfg when the lncli setmccfg is called to change the estimator value. 
  func (s *server) UpdateEstimatorValue(estimator routing.Estimator) {
-	s.cfg.Estimator = estimator
+	switch c := estimator.Config().(type) {
+	case *routing.AprioriConfig:
```
Comment (Chinwendu20) [inline server.go:1690]: I do not think `Config()` returns a pointer hence this should not be a pointer

File: server.go [hunk 7]
```diff
@@ -1686,7 +1686,20 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
...
+		targetCfg.PenaltyHalfLife = c.PenaltyHalfLife
+		targetCfg.Weight = c.AprioriWeight
+		targetCfg.CapacityFraction = c.CapacityFraction
+		targetCfg.HopProbability = c.AprioriHopProbability
+
+	case *routing.BimodalConfig:
+		targetCfg := s.cfg.SubRPCServers.RouterRPC.BimodalConfig
+		targetCfg.Scale = int64(c.BimodalScaleMsat)
+		targetCfg.NodeWeight = c.BimodalNodeWeight
+		targetCfg.DecayTime = c.BimodalDecayTime
```
Comment (Chinwendu20) [inline server.go:1701]: I believe this is missing for the updates to be complete:

```go
		targetCfg.ProbabilityEstimatorType =
			routing.AprioriEstimatorName
```

**Review by Chinwendu20 — COMMENTED** [review]
Nice work. I do think the commit does not follow the guidelines:
https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#ideal-git-commit-structure

File: server.go (hunk 7 above)
Comment (MPins) [reply server.go:1701]: There is no ProbabilityEstimatorType on targetCfg!!!

Comment (Chinwendu20) [reply server.go]: Correct. It should be this:

```go
s.cfg.SubRPCServers.RouterRPC.RoutingConfig.ProbabilityEstimatorType =
			routing.AprioriEstimatorName
```

Comment (MPins) [reply server.go]: Perfect ... it is working now! You were right about the [pointer] (https://github.com/lightningnetwork/lnd/pull/8857#discussion_r1663641513)

**Review by alexbosworth — APPROVED** [review]
utconceptack

**Comment by guggero** [comment]
Can you rebase the PR to get rid of the merge commits please? Then I'll take another look.

**Comment by MPins** [comment]
> Can you rebase the PR to get rid of the merge commits please? Then I'll take another look.

Sorry about that mess! I'm learning a lot of things at the same time and the git collaboration process is one of them! 

Rebase done, thank you for pointing it.

**Comment by guggero** [comment]
Thanks for the rebase. Looks a bit better now.
Unfortunately there's another bit of learning ahead of you. Open Source is quite demanding in terms of git proficiency, since crafting a Pull Request is often as much about the reviewer as it is about the code change itself.
So what you'll want to do to make sure your PRs are getting accepted quickly, is what I call "Review Oriented Programming". Basically you want to tell a coherent story to the reviewer with each commit.

That means each commit should be a small but internally consistent (e.g. each commit compiles on its own) change, with a clear description (e.g. in the commit message or code comments) of why the change is needed and why it's done in this specific way. And each commit should be in a logical order that makes sense to the reviewer.

What you want to avoid is that the commits show different iterations of things you tried or changes during the review itself. So that a new reviewer looking at the code doesn't have to look at all the different stages the code went through and can just take in the current state.

Don't worry if you don't get that perfect for the first couple of pull requests, great commit crafting is a hard skill to learn and an art form in itself.

So long story short: can you please update your PR so it only has 3 or 4 commits? You'll probably want to use interactive rebase to achieve that.

Here are a couple of links that should get you started:
 - https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#ideal-git-commit-structure
 - https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History
 - https://hackernoon.com/beginners-guide-to-interactive-rebasing-346a3f9c3a6d

**Comment by MPins** [comment]
> Thanks for the rebase. Looks a bit better now. Unfortunately there's another bit of learning ahead of you. Open Source is quite demanding in terms of git proficiency, since crafting a Pull Request is often as much about the reviewer as it is about the code change itself. So what you'll want to do to make sure your PRs are getting accepted quickly, is what I call "Review Oriented Programming". Basically you want to tell a coherent story to the reviewer with each commit.
> 
> That means each commit should be a small but internally consistent (e.g. each commit compiles on its own) change, with a clear description (e.g. in the commit message or code comments) of why the change is needed and why it's done in this specific way. And each commit should be in a logical order that makes sense to the reviewer.
> 
> What you want to avoid is that the commits show different iterations of things you tried or changes during the review itself. So that a new reviewer looking at the code doesn't have to look at all the different stages the code went through and can just take in the current state.
> 
> Don't worry if you don't get that perfect for the first couple of pull requests, great commit crafting is a hard skill to learn and an art form in itself.
> 
> So long story short: can you please update your PR so it only has 3 or 4 commits? You'll probably want to use interactive rebase to achieve that.
> 
> Here are a couple of links that should get you started:
> 
> * https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#ideal-git-commit-structure
> * https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History
> * https://hackernoon.com/beginners-guide-to-interactive-rebasing-346a3f9c3a6d

Thank you very much for investing time in providing such valuable guidance. I'll be working on it today and let you know when ready.

**Comment by MPins** [comment]
> Thanks for the rebase. Looks a bit better now. Unfortunately there's another bit of learning ahead of you. Open Source is quite demanding in terms of git proficiency, since crafting a Pull Request is often as much about the reviewer as it is about the code change itself. So what you'll want to do to make sure your PRs are getting accepted quickly, is what I call "Review Oriented Programming". Basically you want to tell a coherent story to the reviewer with each commit.
> 
> That means each commit should be a small but internally consistent (e.g. each commit compiles on its own) change, with a clear description (e.g. in the commit message or code comments) of why the change is needed and why it's done in this specific way. And each commit should be in a logical order that makes sense to the reviewer.
> 
> What you want to avoid is that the commits show different iterations of things you tried or changes during the review itself. So that a new reviewer looking at the code doesn't have to look at all the different stages the code went through and can just take in the current state.
> 
> Don't worry if you don't get that perfect for the first couple of pull requests, great commit crafting is a hard skill to learn and an art form in itself.
> 
> So long story short: can you please update your PR so it only has 3 or 4 commits? You'll probably want to use interactive rebase to achieve that.
> 
> Here are a couple of links that should get you started:
> 
> * https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#ideal-git-commit-structure
> * https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History
> * https://hackernoon.com/beginners-guide-to-interactive-rebasing-346a3f9c3a6d

Done ... only 3 commits!

File: server.go [hunk 8]
```diff
@@ -1680,6 +1683,29 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
 	return s, nil
 }
 
+// UpdateEstimatorValue is a callback function to change estimator value in the  
+// main cfg when the lncli setmccfg is called to change the estimator value. 
+ func (s *server) UpdateEstimatorValue(estimator routing.Estimator) {
+	switch c := estimator.Config().(type) {
+	case routing.AprioriConfig:
+		s.cfg.SubRPCServers.RouterRPC.RoutingConfig.ProbabilityEstimatorType =
```
Comment (guggero) [inline server.go:1691]: This line is very long and the line break after the equal sign looks very weird.
How about:

```
	routerCfg := s.cfg.SubRPCServers.RouterRPC

	switch c := estimator.Config().(type) {
	case routing.AprioriConfig:
		routerCfg.ProbabilityEstimatorType =
			routing.AprioriEstimatorName

		targetCfg := routerCfg.AprioriConfig
		targetCfg.PenaltyHalfLife = c.PenaltyHalfLife
		targetCfg.Weight = c.AprioriWeight
		targetCfg.CapacityFraction = c.CapacityFraction
		targetCfg.HopProbability = c.AprioriHopProbability

	case routing.BimodalConfig:
		routerCfg.ProbabilityEstimatorType =
			routing.BimodalEstimatorName
		targetCfg := routerCfg.BimodalConfig
		targetCfg.Scale = int64(c.BimodalScaleMsat)
		targetCfg.NodeWeight = c.BimodalNodeWeight
		targetCfg.DecayTime = c.BimodalDecayTime
	}
	
```

Also, you seem to be mixing spaces and tabs. Make sure you run `make fmt` on each commit.

**Review by guggero — COMMENTED** [review]
Code and commit structure looks pretty good now! Mostly nits remaining.

File: routing/missioncontrol.go [hunk 9]
```diff
@@ -208,7 +217,7 @@ type paymentResult struct {
 // NewMissionControl returns a new instance of missionControl.
 func NewMissionControl(db kvdb.Backend, self route.Vertex,
 	cfg *MissionControlConfig) (*MissionControl, error) {
-
+
```
[the remark this answers is not in the document]
Comment (MPins) [reply routing/missioncontrol.go:220]: I'll take out the tab ... but don't know how to run `make fmt` yet

[the remark this answers is not in the document]
Comment (guggero) [reply routing/missioncontrol.go:220]: I'm not sure what you mean... It's literally typing the command `make fmt` in your terminal when you're in the `lnd` repository. The same way you run `make build` or `make install`. Or do you get an error running that?
See [here](https://github.com/lightningnetwork/lnd/blob/master/docs/INSTALL.md#building-a-development-version-from-source) for some installation instructions.

[the remark this answers is not in the document]
Comment (MPins) [reply routing/missioncontrol.go:220]: got an error ... make: *** [Makefile:291: fmt] Erro 127

[the remark this answers is not in the document]
Comment (MPins) [reply routing/missioncontrol.go]: Anyway ... I will investigate that tomorrow morning

File: server.go (hunk 8 above)
Comment (MPins) [reply server.go]: I did the changes ... but a little mess on commits are back ... tomorrow morning I will fix it.

Comment (MPins) [reply server.go]: I did just run the `make fmt` ... and fixed the commits ... Now I have done everything you pointed out. Thanks!

**Comment by MPins** [comment]
@guggero nits fixed ... `make fmt` working properly ... commits fixed again. Thanks!

File: routing/missioncontrol.go [hunk 10]
```diff
@@ -208,7 +216,6 @@ type paymentResult struct {
 // NewMissionControl returns a new instance of missionControl.
 func NewMissionControl(db kvdb.Backend, self route.Vertex,
 	cfg *MissionControlConfig) (*MissionControl, error) {
-
```
Comment (guggero) [inline routing/missioncontrol.go:211]: nit: don't delete this empty line please.

File: routing/missioncontrol.go [hunk 11]
```diff
@@ -224,11 +231,12 @@ func NewMissionControl(db kvdb.Backend, self route.Vertex,
 	}
 
 	mc := &MissionControl{
-		state:     newMissionControlState(cfg.MinFailureRelaxInterval),
-		now:       time.Now,
-		selfNode:  self,
-		store:     store,
-		estimator: cfg.Estimator,
+		state:           newMissionControlState(cfg.MinFailureRelaxInterval),
```
Comment (guggero) [inline routing/missioncontrol.go:234]: This now needs to be formatted differently because it exceeds the allowed line length.
For example:
```
	mc := &MissionControl{
		state: newMissionControlState(
			cfg.MinFailureRelaxInterval,
		),
		now:             time.Now,
		selfNode:        self,
		store:           store,
		estimator:       cfg.Estimator,
		UpdateEstimator: cfg.UpdateEstimator,
	}
```

File: routing/missioncontrol.go [hunk 12]
```diff
@@ -93,6 +93,10 @@ type NodeResults map[route.Vertex]TimedPairResult
 // since the last failure is used to estimate a success probability that is fed
 // into the path finding process for subsequent payment attempts.
 type MissionControl struct {
+	// UpdateEstimator is a function that is called whenever the
+	// mission control state is updated.
+	UpdateEstimator func(estimator Estimator)
```
Comment (guggero) [inline routing/missioncontrol.go:98]: This commit needs to come first, otherwise things won't compile (the commit that's currently the first one already references this new field but it's not yet available).

**Review by guggero — COMMENTED** [review]
Tested everything and it works as expected.

A couple of small things to make the linter and other CI steps happy.

File: routing/missioncontrol.go (hunk 12 above)
Comment (MPins) [reply routing/missioncontrol.go:98]: Thank you! Done!

Comment (guggero) [inline routing/missioncontrol.go:98]: nit: since this field only represents the internal state of the `MissionControl` struct and doesn't need to be set or read outside of this file, it doesn't need to be exported (lowercase vs. uppercase in Golang). So can just be `updateEstimator`.

File: routing/missioncontrol.go [hunk 13]
```diff
@@ -121,6 +125,10 @@ type MissionControl struct {
 // MissionControlConfig defines parameters that control mission control
 // behaviour.
 type MissionControlConfig struct {
+	// UpdateEstimatorValue is a function that is called whenever the
```
Comment (guggero) [inline routing/missioncontrol.go:128]: nit: Godoc comment doesn't match struct member name.

**Review by guggero — COMMENTED** [review]
Final nits, I promise :sweat_smile:

Also, you might want to add a short release notes text to the `docs/release-notes/release-notes-0.18.3.md` file (including your name in the list of contributors).
If you rebase onto the latest version of the `master` branch, that file should already be there.

**Comment by MPins** [comment]
> Final nits, I promise 😅
> 
> Also, you might want to add a short release notes text to the `docs/release-notes/release-notes-0.18.3.md` file (including your name in the list of contributors). If you rebase onto the latest version of the `master` branch, that file should already be there.

The process was very important to me. I learned a lot. Thank you. I'll be much more detailed on the following PRs.  😅

File: routing/missioncontrol.go [hunk 14]
```diff
@@ -93,6 +93,10 @@ type NodeResults map[route.Vertex]TimedPairResult
 // since the last failure is used to estimate a success probability that is fed
 // into the path finding process for subsequent payment attempts.
 type MissionControl struct {
+	// UpdateEstimator is a function that is called whenever the
```
Comment (guggero) [inline routing/missioncontrol.go:96]: nit: Godoc should match now lowercase name.

**Review by guggero — APPROVED** [review]
tACK, LGTM :tada:

**Comment by MPins** [comment]
Hello @Abdulkbk and @Chinwendu20 It would be great if you have time to review this PR once more time, I believe it will be the final review. Thanks!

**Review by Abdulkbk — APPROVED** [review]
LGTM

**Review by Abdulkbk — APPROVED** [review]
LGTM

File: routing/missioncontrol.go [hunk 15]
```diff
@@ -121,6 +125,10 @@ type MissionControl struct {
 // MissionControlConfig defines parameters that control mission control
 // behaviour.
 type MissionControlConfig struct {
+	// UpdateEstimator is a callback function that is called whenever the
+	// mission control state is updated.
+	UpdateEstimator func(estimator Estimator)
```
Comment (ellemouton) [inline routing/missioncontrol.go:130]: what about other config values that are not estimator related such as `MaxMcHistory` and `MinFailureRelaxInterval`? 

I think perhaps we should instead make this more generic: `OnConfigUpdate`?

File: server.go [hunk 16]
```diff
@@ -1699,6 +1703,33 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
...
+// values in the main cfg when the lncli setmccfg is called.
+func (s *server) UpdateRoutingConfig(estimator routing.Estimator) {
+	routerCfg := s.cfg.SubRPCServers.RouterRPC
+
+	switch c := estimator.Config().(type) {
+	case routing.AprioriConfig:
+		routerCfg.ProbabilityEstimatorType =
+			routing.AprioriEstimatorName
+
+		targetCfg := s.cfg.SubRPCServers.RouterRPC.AprioriConfig
```
Comment (ellemouton) [inline server.go:1716]: nit: `targetCfg := routerCfg.AprioriConfig`

same below

File: server.go [hunk 17]
```diff
@@ -928,7 +928,11 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
 		}
 	}
 
+	// UpdateRoutingConfig is a function that will be called by the
+	// mission control instance every time a new estimator value
+	// is setted.
```
Comment (ellemouton) [inline server.go:933]: dont think we need this comment here

File: server.go [hunk 18]
```diff
@@ -1699,6 +1703,33 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
 	return s, nil
 }
 
+// UpdateRoutingConfig is a callback function to update the routing config
+// values in the main cfg when the lncli setmccfg is called.
```
Comment (ellemouton) [inline server.go:1707]: nit: it will be called on `SetMissionControlConfig` which is not limited to `lncli setmcfg`

File: rpcserver.go [hunk 19]
```diff
@@ -6920,6 +6920,10 @@ func (r *rpcServer) DebugLevel(ctx context.Context,
 		return nil, err
 	}
 
+	// There is no error parsing the LevelSpec, so we'll update
+	// the current debug level, without checking everything again.
```
Comment (ellemouton) [inline rpcserver.go:6924]: nit: `Propagate the new config level to the main config struct` instead?

File: docs/release-notes/release-notes-0.18.3.md [hunk 20]
```diff
@@ -33,6 +33,11 @@
   created](https://github.com/lightningnetwork/lnd/pull/8921) when multiple RPC
   calls are made concurrently.
 
+* [Fixed a bug](https://github.com/lightningnetwork/lnd/pull/8857) that caused
+  LND not persisting on main cfg the changes done calling the commands 
+  `lncli debuglevel` to change the debuglevel and `lncli setmccfg` to change 
+  the estimator.
```
Comment (ellemouton) [inline docs/release-notes/release-notes-0.18.3.md:39]: nit: "Correctly propagate mission control and debug level config values to the main LND config struct so that the `GetDebugInfo` response is accurate"

File: routing/missioncontrol.go [hunk 21]
```diff
@@ -93,6 +93,10 @@ type NodeResults map[route.Vertex]TimedPairResult
 // since the last failure is used to estimate a success probability that is fed
 // into the path finding process for subsequent payment attempts.
```
Comment (ellemouton) [inline routing/missioncontrol.go:94]: nit: the commit message mentions "persistance" which in my mind means writing to disk. So i'd use "propagate" instead

**Review by ellemouton — CHANGES_REQUESTED** [review]
nice work! My main comment is about ensuring that we account for all MC config values here and not just the estimator

File: routing/missioncontrol.go [hunk 22]
```diff
@@ -308,6 +319,9 @@ func (m *MissionControl) SetConfig(cfg *MissionControlConfig) error {
 	m.state.minFailureRelaxInterval = cfg.MinFailureRelaxInterval
 	m.estimator = cfg.Estimator
 
+	// Callback function to update the estimator value on main cfg.
```
[the remark this answers is not in the document]
Comment (MPins) [reply routing/missioncontrol.go:322]: I thought it would be better for the understanding, without that the comment would be something like "Execute the callback function." ... or maybe it does not need any comment!

File: routing/missioncontrol.go [hunk 23]
```diff
@@ -308,6 +319,9 @@ func (m *MissionControl) SetConfig(cfg *MissionControlConfig) error {
...
+	m.updateEstimator(cfg.Estimator)
```
[the remark this answers is not in the document]
Comment (MPins) [reply routing/missioncontrol.go:323]: I'll be addressing the nil check ... thanks!

**Comment by MPins** [comment]
> nice work! My main comment is about ensuring that we account for all MC config values here and not just the estimator

Thank you for your careful review. I'll be addressing it all shortly.

File: routing/missioncontrol.go [hunk 24]
```diff
@@ -93,6 +93,10 @@ type NodeResults map[route.Vertex]TimedPairResult
 // since the last failure is used to estimate a success probability that is fed
 // into the path finding process for subsequent payment attempts.
 type MissionControl struct {
+	// updateEstimator is a function that is called whenever the
+	// mission control state is updated.
+	updateEstimator func(estimator Estimator)
```
[the remark this answers is not in the document]
Comment (MPins) [reply routing/missioncontrol.go:98]: Maybe onConfigUpdate still better! As you comment below!

**Comment by ellemouton** [comment]
@MPins  - thanks for the quick updates!

I've unresolved one comment cause it is unaddressed: 

> what about other config values that are not estimator related such as MaxMcHistory and MinFailureRelaxInterval? I think perhaps we should instead make this more generic: OnConfigUpdate?

I didnt just mean the name of the call-back. I meant the actual behaviour too. It should take in the `cfg`, not just the `Estimator`. Ie, if I update _any_ MC config  values via the rpc, those should be reflected in `GetDebugInfo`. 



This also needs a rebase as there is a merge conflict

File: routing/missioncontrol.go [hunk 25]
```diff
@@ -110,6 +110,10 @@ type MissionControl struct {
 	// results that mission control collects.
 	estimator Estimator
 
+	// onConfigUpdate is a function that is called whenever the
+	// mission control state is updated.
+	onConfigUpdate func(estimator Estimator)
```
Comment (ellemouton) [inline routing/missioncontrol.go:115]: see my suggestion on the previous review about using `fn.Option` here instead

Comment (MPins) [reply routing/missioncontrol.go:115]: I saw it, on my understanding you proposed using `fn.Option` or a `nil` check, then I have implemented the `nil` check. Maybe I missed somehing.

Comment (ellemouton) [reply routing/missioncontrol.go:115]: I think `fn.Option` is slightly nicer cause then it never lets us forget the nil check :)

File: routing/missioncontrol.go (hunk 15 above)
Comment (MPins) [reply routing/missioncontrol.go]: I did the change. But I did not find `MinFailureRelaxInterval` on main cfg. Maybe I'm not looking for on the right place?

File: routing/missioncontrol.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (MPins) [reply routing/missioncontrol.go]: I'm doing the following:
`onConfigUpdate fn.Option[func(cfg *MissionControlConfig)]`
and Then:
m.onConfigUpdate.WhenSome(func(f func(cfg *MissionControlConfig)) {
		f(cfg.Estimator)
})`
But I'm facing a problem to atribute tthe received function at NewMissionControl function:
`OnConfigUpdate func(cfg *MissionControlConfig)`
to the onConfigUpdate, at here:
`	mc := &MissionControl{
		state: newMissionControlState(
			cfg.MinFailureRelaxInterval,
		),
		now:            time.Now,
		selfNode:       self,
		store:          store,
		estimator:      cfg.Estimator,
		onConfigUpdate: cfg.OnConfigUpdate,
`
`

File: routing/missioncontrol.go (hunk 25 above)
Comment (MPins) [reply routing/missioncontrol.go]: I'm doing the following:
`onConfigUpdate fn.Option[func(cfg *MissionControlConfig)]`
and Then:
`m.onConfigUpdate.WhenSome(func(f func(cfg *MissionControlConfig)) {
		f(cfg.Estimator)
})`
But I'm facing a problem to atribute tthe received function at NewMissionControl function:
`OnConfigUpdate func(cfg *MissionControlConfig)`
to the onConfigUpdate, at here:
`	mc := &MissionControl{
		state: newMissionControlState(
			cfg.MinFailureRelaxInterval,
		),
		now:            time.Now,
		selfNode:       self,
		store:          store,
		estimator:      cfg.Estimator,
		onConfigUpdate: cfg.OnConfigUpdate,
`
cannot use cfg.OnConfigUpdate (variable of type func(cfg *MissionControlConfig)) as fn.Option[func(cfg *MissionControlConfig)] value in struct literalcompiler[IncompatibleAssign](https://pkg.go.dev/golang.org/x/tools/internal/typesinternal#IncompatibleAssign)

File: routing/missioncontrol.go [hunk 26]
```diff
@@ -308,6 +320,11 @@ func (m *MissionControl) SetConfig(cfg *MissionControlConfig) error {
 	m.state.minFailureRelaxInterval = cfg.MinFailureRelaxInterval
 	m.estimator = cfg.Estimator
 
+	// Execute the callback function if it is set.
+	m.onConfigUpdate.WhenSome(func(f func(cfg *MissionControlConfig)) {
+		f(cfg)
+	})
+
```
Comment (ellemouton) [inline routing/missioncontrol.go:327]: 🙏

File: routing/missioncontrol.go [hunk 27]
```diff
@@ -124,6 +129,10 @@ type MissionControlConfig struct {
 	// Estimator gives probability estimates for node pairs.
 	Estimator Estimator
 
+	// OnConfigUpdate is function that is called whenever the
+	// mission control state is updated.
+	OnConfigUpdate func(cfg *MissionControlConfig)
+
```
Comment (ellemouton) [inline routing/missioncontrol.go:135]: this also needs to be `fn.Option[func(cfg *MissionControlConfig)]` though.

Cause currently, if this is nil (which it is right now) then below when you do `onConfigUpdate: fn.Some(cfg.OnConfigUpdate),`, you are setting the callback to a nil. meaning that
```
m.onConfigUpdate.WhenSome(func(f func(cfg *MissionControlConfig)) {
		f(cfg)
	})
``` 
will be executed & cause a panic. 

You can see this by running this commit & calling `lncli setmccfg ...` -> it will cause LND to panic & crash

File: routing/missioncontrol.go (hunk 15 above)
Comment (ellemouton) [reply routing/missioncontrol.go]:
```
	mcCfg := &routing.MissionControlConfig{
		MaxMcHistory: int(req.Config.MaximumPaymentResults),
		MinFailureRelaxInterval: time.Duration(
			req.Config.MinimumFailureRelaxInterval,
		) * time.Second,
	}
```

**Review by ellemouton — COMMENTED** [review]
almost there 🙏

File: routing/missioncontrol.go (hunk 27 above)
Comment (MPins) [reply routing/missioncontrol.go:135]: Sorry, I thought `fn` would take care of it (it is new to me) ... but I should test it.

File: routing/missioncontrol.go (hunk 15 above)
Comment (MPins) [reply routing/missioncontrol.go]: I mean in the target structure `routerrpc.Config`  there is no MinFailureRelaxInterval

Comment (ellemouton) [reply routing/missioncontrol.go]: oh i see 👍

**Comment by ellemouton** [comment]
@MPins  can you rebase & add release notes to 0.19 in the mean time so that the CI can run pls 🙏

**Comment by MPins** [comment]
> @MPins can you rebase & add release notes to 0.19 in the mean time so that the CI can run pls 🙏

Thank you. Done.

**Review by ellemouton — APPROVED** [review]
Thanks! 🛠️
